ledger/src/blocktree.rs

```diff
@@ -1247,27 +1247,30 @@ impl Blocktree {
         slot_duration: Duration,
         stakes: &HashMap<Pubkey, (u64, Account)>,
     ) -> Option<UnixTimestamp> {
-        let mut total_stake = 0;
-        let stake_weighted_timestamps_sum: u64 = self
+        let mut stake_weighted_timestamps: HashMap<Pubkey, (u64, u64)> = HashMap::new();
+        for query_slot in self
             .get_timestamp_slots(slot, TIMESTAMP_SLOT_INTERVAL, TIMESTAMP_SLOT_RANGE)
```

Comment:
Both the original code and updated code here makes my head spin. The original code was more functional (which is usually good), except that it got ugly when it used `sum()` and a mut `total_stake` instead of a `fold()`.  Also, it looks like edge cases weren't very well handled. For example, looks like `get_block_timestamps()` might return an error if the slot is bad, but that won't cause this function to return an error or even None.

In the new code, there's a bunch of redundant calls to `iter()` that cause you to have to dereference double-refs. And I don't think that `HashMap` is actually needed. Why not add to `total_stake` and `stake_weighted_timestamps_sum` where you call `insert()`?

Overall, this code would likely benefit from being broken up into several small functions where each set of edge cases could be addressed independently.

Maybe it'd end up looking something like this:

```rust
let (stake_weighted_timestamps_sum, total_stake) = self
    .get_block_timestamps()?
    .into_iter()
    .filter_map(|x| get_timestamp_and_stake(x, &stakes))
    .fold(|(timestamps, stakes) (timestamp, stake)| (timestamps + timestamp, stakes + stake));
Ok(stake_weighted_timestamps_sum / total_stake)
```

Reply:
Ah, thanks, I did forget about `fold()`.

The need for the HashMap is that the duplications can potentially occur across Items returned by `get_block_timestamps`. I used `insert()` because I want to overwrite an entry in the case of a duplicate. Simply folding them all will result in validators who send multiple timestamped votes in the range in being doubly stake-weighted.

I'll definitely work to improve this code, and I can certainly move the deduping into a separate function, but so far I haven't thought of a great alternative to the HashMap.

Reply:
Updated. @garious , thoughts?
`HashMap::collect()` at least makes the deduplication simpler.
I tried a separate `get_timestamp_and_stake()` method, but I didn't love it, as it needed 6 parameters and actually seemed less readable than these 2 lines of code.